src/hotspot/share/nmt/regionsTree.hpp

```diff
@@ -0,0 +1,174 @@
+  SummaryDiff uncommit_region(address addr, size_t size);
+
+  using Node = VMATree::TreapNode;
+
+  class NodeHelper {
+      Node* _node;
+      public:
+      NodeHelper() : _node(nullptr) { }
+      NodeHelper(Node* node) : _node(node) { }
+      inline bool is_valid() { return _node != nullptr; }
+      inline void clear_node() { _node = nullptr; }
+      inline VMATree::position position() { return _node->key(); }
+      inline bool is_committed_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Committed) >= 2; }
+      inline bool is_released_begin() { return out_state() == VMATree::StateType::Released; }
+      inline bool is_reserved_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Reserved) == 1; }
+      inline VMATree::StateType in_state() { return _node->val().in.type(); }
+      inline VMATree::StateType out_state() { return _node->val().out.type(); }
+      inline size_t distance_from(NodeHelper& other) { return position() - other.position(); }
+      inline NativeCallStackStorage::StackIndex out_stack_index() { return _node->val().out.stack(); }
+      inline MemTag in_flag() { return _node->val().in.mem_tag(); }
+      inline MemTag out_flag() { return _node->val().out.mem_tag(); }
+      inline void set_in_flag(MemTag flag) { _node->val().in.set_flag(flag); }
+      inline void set_out_flag(MemTag flag) { _node->val().out.set_flag(flag); }
+      inline void print_on(outputStream* st) {
+        auto st_str = [&](int s){
+          return s == (int)VMATree::StateType::Released ? "Rl" :
+                 s ==  (int)VMATree::StateType::Reserved ? "Rv" : "Cm";
+        };
+        st->print_cr("pos: " INTPTR_FORMAT " "
+                     "%s, %s <|> %s, %s",
+                     p2i((address)position()),
+                     st_str((int)in_state()),
+                     NMTUtil::tag_to_name(in_flag()),
+                     st_str((int)out_state()),
+                     NMTUtil::tag_to_name(out_flag())
+                     );
+      }
+    };
+
+  void print_on(outputStream* st) {
+    visit_in_order([&](Node* node) {
+      NodeHelper curr(node);
+      curr.print_on(st);
+      return true;
+    });
+  }
+
+  template<typename F> // F == bool(*)(CommittedMemoryRegion&)
+  void visit_committed_regions(const ReservedMemoryRegion& rgn, F func) {
+    size_t start = (size_t)rgn.base();
+    size_t end = (size_t)rgn.end();
+    size_t comm_size = 0;
+    size_t base = start;
+
+    NodeHelper prev;
+    visit_range_in_order(start, end, [&](Node* node) {
+      NodeHelper curr(node);
+      if (prev.is_valid()) {
+        base = prev.position();
+        if (prev.is_committed_begin()) {
+          comm_size += curr.distance_from(prev);
+          if (!curr.is_committed_begin()) {
+            auto st = stack(curr);
+            CommittedMemoryRegion cmr((address)base, comm_size, st);
+            comm_size = 0;
+            prev.clear_node();
```

Comment:
This is unneeded.

Reply:
Which part? Why?
`clear_node()` is the same as `prev = nullptr` if pointers are used.
`is_valid()` checks at each iteration that is it null or not.

Reply:
Because you immediately reassign `prev` to `curr` in L114 it's not necessary to first call `clear_node`.